### PR TITLE
基本支持RWKV6

### PR DESCRIPTION
**Pull request by YuChuXi**
基本支持RWKV6  
转换，加载，lora好了，但是计算图有问题，rwkv_graph.inc:348行附近怎么都改不好，我打了感叹号标记

### CODE REVIEW

**Comment by saharNooby** [comment]
Hi!

Although I've stepped down as the maintainer of `rwkv.cpp`, I think my input will still be valuable:

1. `rwkv_att_v6` looks like a naive (as in "creates and operates on full-blown matrices") implementation of the attention. If it's true, then v6 models will be too slow and probably unusable on large context lengths. [See comparisons for v5](https://github.com/RWKV/rwkv.cpp/pull/148). I recommend pulling an optimized implementation from somewhere and adapting it into `rwkv.cpp`, as was done with v5.
2. there are no tests for v6 models, no Tiny RWKV trained, etc. I would not merge v6 support without proper tests added, because it is asking for something to be unknowingly broken.

**Comment by LaylBongers** [comment]
Hey there! Thanks for the pull request for the long-awaited RWKV V6 support. I'll soon get to reviewing the code, it's on my schedule! Unfortunately I cannot read Chinese, so I've had to get some help translating.
As already mentioned, to make sure support is (and continues to be) functional, we need to include some tests.

**Comment by YuChuXi** [comment]
Sorry, it's not convenient for me to reply at school. 
This pr only supports the conversion and loading of RWKV6. 
The file "rwkv_graph.inc" has an issue near lines 342-347 that I cannot solve.
No matter how I modify it, I may encounter errors such as "GGML_ASSERT: /media/yuchuxi/YuZi/Project/Mozi/rwkv.cpp/ggml/src/ggml.c:6499: ggml_is_contiguous(a)
"

**Comment by YuChuXi** [comment]
I'll go find a small model to test

File: rwkv_graph.inc [hunk 1]
```diff
@@ -320,6 +303,185 @@ static struct ggml_tensor * rwkv_att_v5(
...
+                ctx,
+                ggml_mul_mat(ctx,layer.att_time_decay_w1, xw)
+            )
+        ),
+        layer.att_time_decay
+    );
+    w = ggml_reshape_4d(ctx, w, 1, head_size, head_count, sequence_length);
+
+    // w = torch.exp(-torch.exp(w))
+    w = rwkv_exp(ctx,rwkv_1_minus_x(ctx,rwkv_exp(ctx, w)));
```
Comment (NoahBPeterson) [inline rwkv_graph.inc:431]: Does this call to rwkv_1_minus_x() belong here? The python function isn't `w = torch.exp(1 - torch.exp(w))`, so shouldn't it be `w = rwkv_exp (ctx, ggml_mul (ctx, -1, rwkv_exp (ctx, w)));`

Comment (saharNooby) [reply rwkv_graph.inc:431]: `rwkv_1_minus_x` is defined as `1 - x`, `ggml_neg` is defined as `-x`. So the code looks correct to me -- indeed, we want `exp(1 - exp(w))`, not `exp(-exp(w))`.

Comment (NoahBPeterson) [reply rwkv_graph.inc:431]: This refers to the line here: https://github.com/BlinkDL/ChatRWKV/blob/8c7956743703afddd9bbb09ec5fbaf95e5b05227/RWKV_v6_demo.py#L187 

- `w = torch.exp(-torch.exp(w.float()))`

There's no subtraction operation, only negation

Comment (saharNooby) [reply rwkv_graph.inc:431]: Oh, okay. I misread your comment, sorry; looks like you are right :)

**Comment by YuChuXi** [comment]
[rwkv-x060-173m-pile-20240515-ctx4k.pth](https://huggingface.co/BlinkDL/temp-latest-training-models/blob/main/temp/rwkv-x060-173m-pile-20240515-ctx4k.pth)

This is a small model

**Comment by saharNooby** [comment]
RWKV v6 support was merged in [other PR](https://github.com/RWKV/rwkv.cpp/pull/174).

**Pull request closed without merging**